protocol/contracts/core/MultiOtcOffer.sol

```diff
@@ -0,0 +1,187 @@
+        OTCMarketInfo storage _otcMarketInfo = otcMarketInfo[_marketId]; 
+        _otcMarketInfo.fundsOwner = _fundsOwner;
+        emit FundsOwnerSet(_marketId, _fundsOwner);
+    }
+
+    function setOfferPrice(bytes32 _marketId, uint256 _offerPrice) external override onlyElevatedAccess {
+        if (!_otcMarketIds.contains(_marketId)) { revert InvalidMarketId(_marketId); }
+        OTCMarketPriceParams storage marketPriceParam = otcMarketPriceParams[_marketId];
+        if (_offerPrice < marketPriceParam.minValidOfferPrice || _offerPrice > marketPriceParam.maxValidOfferPrice) revert OfferPriceNotValid();
+        marketPriceParam.offerPrice = _offerPrice;
+        emit OfferPriceSet(_marketId, _offerPrice);
+    }
+
+    function setOfferPriceRange(bytes32 _marketId, uint128 _minValidOfferPrice, uint128 _maxValidOfferPrice) external override onlyElevatedAccess {
+        if (_minValidOfferPrice > _maxValidOfferPrice) revert CommonEventsAndErrors.InvalidParam();
+        OTCMarketPriceParams storage marketPriceParam = otcMarketPriceParams[_marketId];
+        marketPriceParam.minValidOfferPrice = _minValidOfferPrice;
+        marketPriceParam.maxValidOfferPrice = _maxValidOfferPrice;
+        emit OfferPriceRangeSet(_marketId, _minValidOfferPrice, _maxValidOfferPrice);
+    }
+
+    /// @notice Owner can pause user swaps from occuring
+    function pause() external override onlyElevatedAccess {
+        _pause();
+    }
+
+    /// @notice Owner can unpause so user swaps can occur
+    function unpause() external override onlyElevatedAccess {
+        _unpause();
+    }
+
+    /// @notice Swap `userSellToken` for `userBuyToken`, at the `offerPrice`
+    function swap(bytes32 marketId, uint256 sellTokenAmount) external override whenNotPaused returns (uint256 buyTokenAmount) {
+        if (!_otcMarketIds.contains(marketId)) { revert InvalidMarketId(marketId); }
+        if (sellTokenAmount == 0) revert CommonEventsAndErrors.ExpectedNonZero();
+        OTCMarketInfo memory marketInfo = otcMarketInfo[marketId];
+        buyTokenAmount = quote(marketId, sellTokenAmount);
+        address _fundsOwner = marketInfo.fundsOwner;
+        IERC20Metadata _userSellToken = marketInfo.userSellToken;
+        IERC20Metadata _userBuyToken = marketInfo.userBuyToken;
+        emit Swap(msg.sender, _fundsOwner, sellTokenAmount, buyTokenAmount);
+
+        _userSellToken.safeTransferFrom(msg.sender, _fundsOwner, sellTokenAmount);
+        _userBuyToken.safeTransferFrom(_fundsOwner, msg.sender, buyTokenAmount);
+    }
+
+    /// @notice How many `userBuyToken` you would receive given an amount of `sellTokenAmount`
+    function quote(bytes32 marketId, uint256 sellTokenAmount) public override view returns (uint256 buyTokenAmount) {
+        if (!_otcMarketIds.contains(marketId)) { revert InvalidMarketId(marketId); }
+        OTCMarketPriceParams memory marketPriceParam = otcMarketPriceParams[marketId];
```

Comment:
should be `storage` imo

Reply:
can you explain more? Because we're not changing state in this view function, just reading.

Reply:
If the memory keyword is used then it does a full copy of the struct from storage into memory upfront. If you use storage keyword then it just acts like a pointer, it doesn't have to do the full copy, just access the bit it needs.

Best to compare in tests though